docs/proposals/train_api_proposal.md

```diff
@@ -0,0 +1,154 @@
+
+```python
+training_client.create_job(
+   name=pytorchjob_name,
+   train_func=train_function,
+   num_worker_replicas=3, # How many PyTorch Workers will be created.
+)
+```
+
+The above method reduces the time taken but still the end user has to write the train_function method.
+
+In many cases like in LLM, training core code template is a standard and it doesn’t require any custom modification. For faster experimentation of such models, an easier api should exist to train/finetune models without writing custom Python code. HuggingFace Trainer class is a good example but it is not easy to set it up in a Kubernetes environment. Training operator can handle this complexity and provide a better abstracted layer to the user.
+
+**<h3>New Train API proposal:</h3>**
+
+```python
+From kubeflow.training import trainingClient
+# Arguments related to the model provider including credentials
+model_args = modelProviderClass()
+# Arguments related to the dataset provider including credentials
+dataset_args = datasetProviderClass()
+# Arguments related to the trainer code
+parameters = {key:value pairs}
+trainingClient.train(
+   nodes=1, 
+   nprocs_per_node=4, 
+   model='provider://repo/model_name', 
+   dataset= 'provider://train_dataset_path',
+   eval_dataset = "provider://eval_dataset_path"
+   model_args, 
+   dataset_args, 
+   parameters
+)
+```
+
+Example: 
+
+```python
+@dataclass
+class HuggingFace:
+    access_token = field()
+    peft_config = field()
+    transformerClass = field()
+
+@dataclass
+class S3:
+    access_token = field()
+    region = field()
+ 
+trainingClient.train(
+   nodes=1, 
+   nprocs_per_node=4, 
+   model='hf://openchat/openchat_3.5', 
+   dataset= 's3://doc-example-bucket1/train_dataset',
+   eval_dataset = "s3://doc-example-bucket1/eval_dataset"
+   HuggingFace(access_token = "hf_..." , peft_config = {lora_alpha: 16}, transformerClass="Trainer"), 
+   s3(access_token = "s3 access token", region="us-west-2"), 
+   {learning_rate=0.1}
+)
+```
+
+The new proposed API takes following arguments 
+
+1. System parameters - Number of nodes, number of procs per node(GPUs per node).
+2. Model parameters - Model provider and repository details.
+3. Dataset parameters - Dataset provider and dataset details.
+4. Training parameters - Training specific parameters like learning rate etc.
+
+**<h3>Implementation</h3>**
+
+1. Setup **init** **containers** that download the model and dataset to a PVC. Based on the specified model provider, corresponding training utility functions will be used. Eg: For Huggingface provider, Huggingface trainer can be used. For this **get_pytorchjob_template** function in the sdk needs to be changed to add init containers spec.. Inorder to download models and data sets, we need to support different providers like kaggle, hugging face, s3 or git lfs. The data can be stored in a shared volume between the init container and the main container.
```

Comment:
>The data can be stored in a shared volume between the init container and the main container.

Are we going to use EmptyDit to share volume ?

Reply:
we can use nfs or host path volume, is there any particular reason to choose Emptydir?

Reply:
@deepanker13 @johnugeorge Should we discuss pros and cons for each approach ?
E.g. KServe users EmptyDir to share volume between Storage Initializer and Serving container: https://github.com/kserve/kserve/blob/829c79e052d70837c9d1cb0e2bd3418471746733/pkg/webhook/admission/pod/storage_initializer_injector.go#L249-L255

Reply:
I think that `PVC` is better because `EmptyDir` is so limited resources as, generally, it is allocated from Node Volumes, excluding the case of using the GenericEphemeralVolume or CSIEphemeralVolumes.

As an alternative, we may be able to switch which volumes (PVC vs GenericEphemeralVolume) we should use based on the provisioning mode supported in the default storage class.

Reply:
@tenzen-y for PVC what type of volume are you suggesting?

Reply:
@tenzen-y Can we investigate this after initial implementation ? Using PVC might be not straightforward since it requires Kubeflow cluster administrators to support `ReadOnlyMany` in their volume provisioners.

Also, in the future we can take a look at KServe model car proposal: https://github.com/kserve/kserve/issues/3043 for model download from the OCI.

Reply:
> @tenzen-y for PVC what type of volume are you suggesting?

My comment was for @andreyvelich's one. Sorry for your confusion.

Reply:
> @tenzen-y Can we investigate this after initial implementation ? Using PVC might be not straightforward since it requires Kubeflow cluster administrators to support ReadOnlyMany in their volume provisioners.
> 
> Also, in the future we can take a look at KServe model car proposal: https://github.com/kserve/kserve/issues/3043 for model download from the OCI.
> 

Yes, sure. We can investigate more carefully in the future. 
In the first implementation, I think we can take EmptyDir way. WDYT?

Reply:
If we do not use PVC, won't it be very inefficient to download large models in every worker pod?

Reply:
> If we do not use PVC, won't it be very inefficient to download large models in every worker pod?

It is correct. It might be better to create batch/v1 Job to download models to PVC. After downloading completed, we can start PyTorchJob.

Reply:
> @johnugeorge @tenzen-y Do we have any limitation for ephemeralStorage on Kubernetes clusters ?
Otherwise, we can investigate [emptyDir.medium](https://kubernetes.io/docs/concepts/storage/volumes/#emptydir) which uses tmpfs and stores data in memory.

I'd like to avoid node resources (system memory, local storage, etc...) since those resources are so limited and expensive. As much as possible, I'd like to store the models in resources (disk/storage) separate from Node, which means PVC.

The separate resources have extensibility and cost efficiency rather than node resources.

Reply:
>It is correct. It might be better to create batch/v1 Job to download models to PVC

This definitely will be more efficient rather than downloading model on every Worker initContainer, but we need to design how we are going to orchestrate such initialization Job, provides guidelines for users that they need to have StorageClass to support `ReadOnlyMany` PVC. 

Should we extend scope of this proposal to cover this as well ?

Reply:
> we need to design how we are going to orchestrate such initialization Job, provides guidelines for users that they need to have StorageClass to support ReadOnlyMany PVC.

This concern makes sense. I agree with extending the scope of this proposal.
Ideally, we should define the way to pull models as CRDs. However, at the first step, we want to support only via SDK without CRD since it would be hard work. Maybe we can implement CRDs later.

It means that I would suggest extending this proposal so that we can design the way to pull models using batch/v1 Job, not CRD.

@andreyvelich @johnugeorge WDYT?

Reply:
Yes, we discussed this with @johnugeorge that we can just download model on the `master` and then attach volume to the `worker` replicas.
I think, we should explain this in the proposal that the volume should support `ReadOnlyMany` mode for this.

Reply:
SGTM